OCPBUGS-11437: MCO keeps the pull secret to .orig file once it replaced

inesqyx opened the pull request:
- OCPBUGS11437
The pull secret is preserved in the path /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig. Whereas, the user needs it to be deleted when the pull secret is being deleted/ replaced.
- What is the problem
There is complaint about that the MCD is creating orig files to preserve files that should not be preserved. To be more specific, the orig preserving system should only preserve files that are originally on the disk, but are now preserving files that are written by Ignition. The root cause is that Ignition writes the files before the MCD takes the control as a result when MCD comes in, it sees all the files written before as files on disk, which is very wrong.
- What I did
Add more constraints for orig file generation: orig files should only be generated on files that 1) claimed by the rpm 2) exist in the usr/etc directory 3) exist before MCD takes control
- How to verify it
Ignition writes files in the /var directory. And now when cat into the orig folder for the MCD, there is no var folder, meaning that files written by Ignition are not preserved now. Specific example would be that catting into /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdori will now return no so such file

inesqyx commented:
/test all

inesqyx commented:
/retest-required

sergiordlr commented:
Verified using IPI on OSP

1) Verify that the file` /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig` does not exist

```
oc debug node/sregidor-nps-zgw74-worker-0-9cc8x -- chroot /host cat /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig
Starting pod/sregidor-nps-zgw74-worker-0-9cc8x-debug-n2zmb ...
To use host binaries, run `chroot /host`
cat: /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig: No such file or directory
```


2) Create a MC that deploys a file and remove the MC

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-machine-config
spec:
  config:
    ignition:
      version: 3.1.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,dGVzdA==
        filesystem: root
        mode: 420
        path: /etc/test-file.test
```

The file is created and removed normally. We can read the delete-stale logs when removing the file
```
I0628 11:44:27.559549    2249 update.go:1250] Deleting stale data
I0628 11:44:27.559664    2249 update.go:1313] Removed stale file "/etc/test-file.test"
```


3) Create a MC that deploy the /var/lib/kubelet/config.json and remove it
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-machine-config
spec:
  config:
    ignition:
      version: 3.1.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,e30=
        filesystem: root
        mode: 0600
        path: /var/lib/kubelet/config.json
```

When we deploy the MC the file /var/lib/kubelet/config.json has this content: {}

When we remove the MC the file  /var/lib/kubelet/config.json has the content defined in the pull-secret.


4) Configure an empty pull-secret: {}

After configuring the emtpy pull-secret the file /var/lib/kubelet/config.json does not exist

```
]$ oc debug node/sregidor-nps-zgw74-worker-0-9cc8x -- chroot /host cat /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig
Starting pod/sregidor-nps-zgw74-worker-0-9cc8x-debug-slf4r ...
To use host binaries, run `chroot /host`
cat: /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig: No such file or directory
```

5) Upgrade from 4.13

Before the upgrade, the file exists and contains the right content:
```
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.nightly-2023-06-20-224158   True        False         15m     Cluster version is 4.13.0-0.nightly-2023-06-20-224158
$ oc debug node/sregidor-up2-7n9q7-worker-0-jbskz -- chroot /host cat /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig | head -c 10
Starting pod/sregidor-up2-7n9q7-worker-0-jbskz-debug-vc8n6 ...
To use host binaries, run `chroot /host`
{"auths":{
Removing debug pod ...
```

After the upgrade, the file still exists

```
 oc debug node/sregidor-up2-7n9q7-worker-0-jbskz -- chroot /host head -c 10 /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig
Starting pod/sregidor-up2-7n9q7-worker-0-jbskz-debug-qh7wf ...
To use host binaries, run `chroot /host`
{"auths":{
Removing debug pod ...
```

Even if we re-configure the pull-secret, and we set an empty pull-secret, the file is not removed.




6) Configure a MC using a file that exists in the machines inside /etc/ and has "config.json" suffix.

Look for that file:

```
Debug a worker node:

sh-5.1# find . -iname "*config.json" 2> /dev/null |grep -v storage
./run/openshift-sdn/cniserver/config.json
./sysroot/ostree/deploy/rhcos/var/lib/kubelet/config.json
./usr/etc/tpm2-tss/fapi-config.json
./var/lib/kubelet/config.json
./var/lib/kubelet/pods/b482e398-27f6-4dac-b8c0-6715c0010f99/volumes/kubernetes.io~secret/installation-pull-secrets/config.json
./var/lib/kubelet/pods/b482e398-27f6-4dac-b8c0-6715c0010f99/volumes/kubernetes.io~secret/installation-pull-secrets/..2023_06_28_13_26_35.4113272500/config.json
./etc/tpm2-tss/fapi-config.json
sh-5.1# cat ./etc/tpm2-tss/fapi-config.json
{
     "profile_name": "P_ECCP256SHA256",
     "profile_dir": "/etc/tpm2-tss/fapi-profiles/",
     "user_dir": "~/.local/share/tpm2-tss/user/keystore",
     "system_dir": "/var/lib/tpm2-tss/system/keystore",
     "tcti": "",
     "system_pcrs" : [],
     "log_dir" : "/run/tpm2-tss/eventlog/"
}
```

We will use /etc/tpm2-tss/fapi-config.json. We create a MC to replace the file in the machine


```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-fapi-config
spec:
  config:
    ignition:
      version: 3.1.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,ewoKCgoKCiAgICAgInByb2ZpbGVfbmFtZSI6ICJQX0VDQ1AyNTZTSEEyNTYiLAogICAgICJwcm9maWxlX2RpciI6ICIvZXRjL3RwbTItdHNzL2ZhcGktcHJvZmlsZXMvIiwKICAgICAidXNlcl9kaXIiOiAifi8ubG9jYWwvc2hhcmUvdHBtMi10c3MvdXNlci9rZXlzdG9yZSIsCiAgICAgInN5c3RlbV9kaXIiOiAiL3Zhci9saWIvdHBtMi10c3Mvc3lzdGVtL2tleXN0b3JlIiwKICAgICAidGN0aSI6ICIiLAogICAgICJzeXN0ZW1fcGNycyIgOiBbXSwKICAgICAibG9nX2RpciIgOiAiL3J1bi90cG0yLXRzcy9ldmVudGxvZy8iCn0KCg==
        filesystem: root
        mode: 0644
        path: /etc/tpm2-tss/fapi-config.json

```

The the configuration is applied without problems.

We see that there is no "orig" file created for this file:
```
sh-5.1# cat ./etc/tpm2-tss/fapi-config.json
{





     "profile_name": "P_ECCP256SHA256",
     "profile_dir": "/etc/tpm2-tss/fapi-profiles/",
     "user_dir": "~/.local/share/tpm2-tss/user/keystore",
     "system_dir": "/var/lib/tpm2-tss/system/keystore",
     "tcti": "",
     "system_pcrs" : [],
     "log_dir" : "/run/tpm2-tss/eventlog/"
}

sh-5.1# find . -iname "*orig" 2> /dev/null |grep fapi
sh-5.1# exit
```

When we remove the MC, the file is deleted. We can see these lines in the daemon logs

```
I0628 13:57:44.872194    2388 update.go:1250] Deleting stale data
I0628 13:57:44.872282    2388 update.go:1313] Removed stale file "/etc/tpm2-tss/fapi-config.json"
```

And the original file is lost:

```
sh-5.1# cat /etc/tpm2-tss/fapi-config.json 
cat: /etc/tpm2-tss/fapi-config.json: No such file or directory
```



My **main concerns** after the verification steps are:

- **When we upgrade a cluster** to a version with this fix, **the file** `/etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig` is still present in the cluster, and **is not removed**.

- **When we try reconfigure any file that is present in the cluster and has the suffix "config.json"**, **the file**'soriginal content is not tracked and **can be eventually deleted**. In the current verification steps we used "/etc/tpm2-tss/fapi-config.json", but since the OS versions will be evolving, in the future it can happen that more files like this one are affected. Maybe the current file is not important, but it is possible that the ones in the future are.


Before adding the qe-appoved label I would like to make sure that those behaviors are OK.

Thank you very much!

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -60,6 +61,12 @@ func createOrigFile(fromPath, fpath string) error {
 		return writeFileAtomicallyWithDefaults(noOrigFileStampName(fpath), nil)
 	}
 
+	// https://issues.redhat.com/browse/OCPBUGS-11437
```
When commenting a bug like this, it means that the code that follows is intended as a workaround, and we would want to remove the code at a later time.

I think that makes sense here, since we eventually want to fix the "orig file" implementation, either separately or eventually as part of layering. Maybe it would be good to mention that here? i.e. this code can be eventually superceded by a proper fix to the underlying problem

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -60,6 +61,12 @@ func createOrigFile(fromPath, fpath string) error {
 
+	// https://issues.redhat.com/browse/OCPBUGS-11437
+	// MCO keeps the pull secret to .orig file once it replaced
+	if strings.HasSuffix(fpath, "config.json") {
```
I think like Sergio mentions, this probably should check specifically for the pull secret path and not just any config.json

inesqyx replied:
Thanks, @sergiordlr and @yuqi-zhang for pointing this out. I have made the changes to more carefully point at the desired path, and yet to be tested.

yuqi-zhang reviewed (commented):
Thanks Ines! The general approach makes sense to me, some comments:

> When we upgrade a cluster to a version with this fix, the file /etc/machine-config-daemon/orig/var/lib/kubelet/config.json.mcdorig is still present in the cluster, and is not removed.

Agree with Sergio. I think eventually we want to remove the orig approach. We could make it such that we delete the orig file entirely for the pull secret even on upgrades as noted by Sergio's test process. I don't think it must be part of this PR though, so it depends on what you want to do @inesqyx 

Also minor nit: if you can update the commit message to also have some of the context in your PR message, that would be good as well!

sergiordlr commented:
I set the jira status to "assigned" again, please, set it back to "post" when the PR is ready for verification again.

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

cgwalters commented:
I think we can more categorically exclude everything in `/var` from the .orig treatment btw, as a big part of the idea here is that `/var` starts out empty on ostree systems.

That would avoid hardcoding anything relating to the pull secret.

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,50 @@ func noOrigFileStampName(fpath string) string {
+	if _, err := exec.Command("rpm", "-qf", fpath).CombinedOutput(); err == nil {
+		// File is owned by an rpm
+		orig = true
+	} else if strings.HasPrefix(fpath, "/etc") && isCoreOSVariant {
```
Hmm, could you remind me why we are explicitly doing this `isCoreOSVariant` check?

inesqyx replied:
The check isCoreVariant checks whether the underlying OS is FCOS or a derivative (including SCOS and RHCOS) that uses ostree + Ignition. These OSs keep the defaults in the /usr/etc directory. As a result, we add the check here to ensure that the file is in the default-keeping dir of the corresponding OS, so it should be orig-ed.

yuqi-zhang replied:
Hmm, ok, I think I am fine either way, but to keep consistent with existing behaviour, maybe it would be best to remove the check. The path is not entirely correct regardless, and I think keeping CoreOS and non-CoreOS nodes consistent may make it easier in the future. What do you think?

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,50 @@ func noOrigFileStampName(fpath string) string {
+	} else if forceCreate {
+		orig = true
+	} else {
+		orig = false
```
nit: you can declare orig as false originally and skip this else condition (i.e. `var orig bool` -> `orig := false`

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,50 @@ func noOrigFileStampName(fpath string) string {
+	// https://issues.redhat.com/browse/OCPBUGS-11437
+	// MCO keeps the pull secret to .orig file once it replaced
+	// Adapt a check used in function deleteStaleData for backwards compatibility: basically if the file doesn't
+	// exist in /usr/etc (on FCOS/RHCOS) and no rpm is claiming it, we assume that the orig file came from a
```
nit: `we assume that the orig file came from a wrongful backup of a MachineConfig file instead of a file originally on disk.` -> `we assume the orig file doesn't need to be created`

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,50 @@ func noOrigFileStampName(fpath string) string {
-		return nil
+func createOrigFile(fromPath, fpath string, isCoreOSVariant, forceCreate bool) error {
+
+	// Declare a boolean variable - orig - to indicate whether an orig file should be created
```
nit: I think this set of comments can be skipped, since the code does a good job explaining what this bool is used for

yuqi-zhang reviewed (commented):
Overall makes sense to me! Just a few nits and a question inline:

yuqi-zhang commented:
/jira refresh

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -1239,7 +1239,9 @@ func (dn *Daemon) isPathInDropins(path string, systemd *ign3types.Systemd) bool
 //
 //nolint:gocyclo
 func (dn *Daemon) deleteStaleData(oldIgnConfig, newIgnConfig ign3types.Config) error {
+
```
nit: there seems to be some extra whitespace added here, would you be able to remove?

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,42 @@ func noOrigFileStampName(fpath string) string {
+		} else {
+			orig = true
+		}
+	} else if forceCreate {
```
nit: since forceCreate always makes `orig` true, it might be better to have this condition check first, to save some extra operations

yuqi-zhang reviewed (commented):
some nits below, functionality should work, @sergiordlr would you be able to verify again? Thanks!

yuqi-zhang commented:
Had a chat with Ines, I think the plan then is to:

1. remove the force flag for the function
2. keep the restoration logic as in the PR (and not only just the pull secret, since other files are also affected)
3. change the tests to test for the non-backup scenario instead, so the force flag isn't required
4. simplify the logic a bit around that

Would that work @sergiordlr ?

sergiordlr commented:
@yuqi-zhang Sure, ofc. Thank you very much!

inesqyx commented:
/retest-required

inesqyx commented:
@sergiordlr Hi, Sergio, I have fixed the nits and updated the test as planned. Both of them and the functionality should be reflected in the latest push (verified locally). Can you please help me verify it? Thank you!

inesqyx commented:
/retest-required

inesqyx commented:
/test e2e-hypershift

cheesesashimi commented on pkg/daemon/file_writers.go:
```diff
@@ -46,11 +47,39 @@ func noOrigFileStampName(fpath string) string {
+	// MCO keeps the pull secret to .orig file once it replaced
+	// Adapt a check used in function deleteStaleData for backwards compatibility: basically if the file doesn't
+	// exist in /usr/etc (on FCOS/RHCOS) and no rpm is claiming it, we assume the orig file doesn't need to be created
+	if _, err := exec.Command("rpm", "-qf", fpath).CombinedOutput(); err == nil {
```
suggestion (non-blocking): Taking a look at this RPM query and its analogue in the `deleteStaleData` function, the code flow seems very similar in both places. Given that, it would be nice improvement to consolidate the RPM query logic into a single function that both `createOrigFile()` and `deleteStaleData()` can use.

cheesesashimi commented on pkg/daemon/update_test.go:
```diff
@@ -774,51 +774,20 @@ func TestOriginalFileBackupRestore(t *testing.T) {
 	err = createOrigFile(controlFile, controlFile)
 	assert.Nil(t, err)
 
-	// Restore the normal file
```
question: Are these tests no longer needed or do they now fail with this new change?

inesqyx replied:
I think I am deleting those because they now fail with this new change.  This is because the tests create files in the /tmp folder, try to create orig files for those and see if that is successful by restoring form the orig files. But the current workaround stops the orig file preservation for the files in the /tmp folder as they are not the original files shipped with the disk, so the tests are no longer valid. However, I have also tested the  current createOrig' s functionality in preserving the files by putting a forceCreate variable in, so that when running the test, the origs are forcedly created for these /tmp files. After verifying its functionality, I removed those variables as well as the tests.

yuqi-zhang replied:
To give a bit more context, I think these tests aren't really needed anymore since we are now (properly) not preserving anything written to /tmp, since they shouldn't have been preserved in the first place.

We could instead test preservation of `/etc` specifically but it is a bit harder to do so via a unit test and may modify the user's system, so we decided to remove this test for now. If there is a simple way to test that that we're missing, would also be happy to take suggestions!

cheesesashimi reviewed (commented):
Overall this looks good! I have a few suggestions and questions I'd like you to consider.

yuqi-zhang commented on pkg/controller/common/helpers.go:
```diff
@@ -1159,7 +1159,6 @@ func (n namespacedEventRecorder) Eventf(object runtime.Object, eventtype, reason
 func (n namespacedEventRecorder) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{}) {
 	n.delegate.AnnotatedEventf(ensureEventNamespace(object), annotations, eventtype, reason, messageFmt, args...)
 }
-
```
Nit: it looks like there is a line removed here, was there a reason for that? If not could you reset this? Thanks!

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,62 @@ func noOrigFileStampName(fpath string) string {
 	return filepath.Join(noOrigParentDir(), fpath+".mcdnoorig")
 }
 
+func isFileOwnedByRPMPkg(fpath string) (string, bool, error) {
```
In both calls to this function, only the middle bool (isowned) is used. Was that how you intended to use this? If so, we can maybe just return the bool.

Also, I think we should always process the error. In the calls to this function the error is ignored. I think we should consider processing the error properly

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,62 @@ func noOrigFileStampName(fpath string) string {
+	if strings.Contains(string(out), fileNotOwnedMsg) {
+		return "", false, nil
+	}
+	return "", false, fmt.Errorf("command %q terminated unexpectedly: %s: %w", cmd, string(out), err)
```
nit: instead of terminated, I think you can just say something like `command %q returned with unexpected error:`

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

inesqyx commented:
/retest-required

inesqyx commented:
/test unit

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,80 @@ func noOrigFileStampName(fpath string) string {
 }
 
+func isFileOwnedByRPMPkg(fpath string) (bool, bool, error) {
+	// The first bool returned indicated whether the underlying OS is Mac
```
nit: for clarity purposes, maybe something like:

`The first bool returns false if rpm exists, and true otherwise (indicating other Linux distros or Mac). We would like to skip orig file creation and preservation`

yuqi-zhang commented on pkg/daemon/file_writers.go:
```diff
@@ -45,12 +46,80 @@ func noOrigFileStampName(fpath string) string {
+			}
+		} else if isMac {
+			// Run on Mac
+			klog.Infof("Running on Mac,skip orig file preservation.")
```
nit: same as above, I think rephrase to "rpm binary not found", and change the references from Mac to "rpmNotFound" or similar, I think that's more clear in case someone gets this error on a non-fedora/rhel machine for some reason

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -1282,20 +1283,27 @@ func (dn *Daemon) deleteStaleData(oldIgnConfig, newIgnConfig ign3types.Config) e
 				}
+			} else if isMac {
+				// Run on Mac
+				klog.Infof("Running on Mac,skip file restoration.")
```
nit: same as above

yuqi-zhang commented on pkg/daemon/update_test.go:
```diff
@@ -817,51 +818,38 @@ func TestOriginalFileBackupRestore(t *testing.T) {
+	assert.Nil(t, errr)
+	assert.Nil(t, errr)
+	// fileNotOwnedMsg := "not owned"
+	// if strings.Contains(string(out), fileNotOwnedMsg) {
```
nit: remove unused code

yuqi-zhang reviewed (commented):
some nits on logs and comments, otherwise lgtm!

yuqi-zhang commented on pkg/daemon/update_test.go:
```diff
@@ -817,51 +818,38 @@ func TestOriginalFileBackupRestore(t *testing.T) {
-	assert.Nil(t, err)
+	// Check whether the given path is owned by an rpm pkg
+	cmd := exec.Command(path, "-qf", path)
+	_, errr := cmd.CombinedOutput()
```
just curious, why is this errr? Why not just `_, err = cmd.CombinedOutput()`

rioliu-rh commented:
@inesqyx @yuqi-zhang any update on this PR? we have a automation PR depends on this one. https://github.com/openshift/openshift-tests-private/pull/11128
/cc @sergiordlr

inesqyx commented:
> @inesqyx @yuqi-zhang any update on this PR? we have a automation PR depends on this one. [openshift/openshift-tests-private#11128](https://github.com/openshift/openshift-tests-private/pull/11128) /cc @sergiordlr

Hey, Rio! I am back from the PTO and sorry for the late reply. I will have the changes in place and make it through in a day or so. Should be quick. Thanks for the reminder!

inesqyx commented:
/retest-required

yuqi-zhang reviewed (approved):
/lgtm

Thanks for the fixes! @rioliu-rh would you say this is a critical fix then? Merging to master (4.14) is currently blocked unless it can be considered a critical fix

inesqyx commented:
/jira refresh
